engine/access/state_stream/backend/backend_events_test.go

```diff
@@ -128,77 +222,206 @@ func (s *BackendEventsSuite) runTestSubscribeEvents() {
+	requireFn func(interface{}, *EventsResponse),
+	tests []eventsTestType,
+) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
 	for _, test := range tests {
 		s.Run(test.name, func() {
-			s.T().Logf("len(s.execDataMap) %d", len(s.execDataMap))
-
 			// add "backfill" block - blocks that are already in the database before the test starts
 			// this simulates a subscription on a past block
-			for i := 0; i <= test.highestBackfill; i++ {
-				s.T().Logf("backfilling block %d", i)
-				s.executionDataTracker.On("GetHighestHeight").
-					Return(s.blocks[i].Header.Height).Maybe()
+			if test.highestBackfill > 0 {
+				s.highestBlockHeader = s.blocks[test.highestBackfill].Header
 			}
 
 			subCtx, subCancel := context.WithCancel(ctx)
-			sub := s.backend.SubscribeEvents(subCtx, test.startBlockID, test.startHeight, test.filters)
 
-			// loop over all of the blocks
+			// mock latest sealed if test case no start value provided
+			if test.startBlockID == flow.ZeroID && test.startHeight == 0 {
+				s.snapshot.On("Head").Unset()
+				s.snapshot.On("Head").Return(s.rootBlock.Header, nil).Once()
+			}
+
+			sub := subscribeFn(subCtx, test.startBlockID, test.startHeight, test.filter)
+
+			// loop over all blocks
 			for i, b := range s.blocks {
-				s.T().Logf("checking block %d %v", i, b.ID())
+				s.T().Logf("checking block %d %v %d", i, b.ID(), b.Header.Height)
 
-				// simulate new exec data received.
-				// exec data for all blocks with index <= highestBackfill were already received
+				// simulate new block received.
+				// all blocks with index <= highestBackfill were already received
 				if i > test.highestBackfill {
-					s.T().Logf("checking block %d %v", i, b.ID())
-
-					s.executionDataTracker.On("GetHighestHeight").Unset()
-					s.executionDataTracker.On("GetHighestHeight").
-						Return(b.Header.Height).Maybe()
+					s.highestBlockHeader = b.Header
 
 					s.broadcaster.Publish()
 				}
 
 				var expectedEvents flow.EventsList
 				for _, event := range s.blockEvents[b.ID()] {
-					if test.filters.Match(event) {
+					if test.filter.Match(event) {
 						expectedEvents = append(expectedEvents, event)
 					}
 				}
 
-				// consume execution data from subscription
+				// consume events response from subscription
 				unittest.RequireReturnsBefore(s.T(), func() {
 					v, ok := <-sub.Channel()
-					require.True(s.T(), ok, "channel closed while waiting for exec data for block %d %v: err: %v", b.Header.Height, b.ID(), sub.Err())
+					require.True(s.T(), ok, "channel closed while waiting for exec data for block %x %v: err: %v", b.Header.Height, b.ID(), sub.Err())
 
-					resp, ok := v.(*EventsResponse)
-					require.True(s.T(), ok, "unexpected response type: %T", v)
+					expected := &EventsResponse{
+						BlockID:        b.ID(),
+						Height:         b.Header.Height,
+						Events:         expectedEvents,
+						BlockTimestamp: b.Header.Timestamp,
```

Comment:
can you include `MessageIndex` here as well?

Reply:
Currently, `MessageIndex` tracking has been relocated to the `handler`, so it isn't directly accessible here in the `backend`.